File: VB/CustomControlInGrid/CustomRepositoryItem/CustomControl.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Controls
Imports System.ComponentModel
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors.ViewInfo

Namespace CustomControlInGrid
	Friend Class CustomControl
		Inherits BaseEdit
		Implements IAutoHeightControl
		#Region "IAutoHeightControl implement"
		Private ReadOnly Property SupportsAutoHeight() As Boolean Implements IAutoHeightControl.SupportsAutoHeight
			Get
				Return True
			End Get
		End Property

		Public Event eHeightChanged As EventHandler
		Private Custom Event HeightChanged As EventHandler Implements IAutoHeightControl.HeightChanged
			AddHandler(ByVal value As EventHandler)
				AddHandler eHeightChanged, value
			End AddHandler
			RemoveHandler(ByVal value As EventHandler)
				RemoveHandler eHeightChanged, value
			End RemoveHandler
			RaiseEvent(ByVal sender As System.Object, ByVal e As System.EventArgs)
			End RaiseEvent
		End Event
		Protected Sub RaiseHeightChanged()
			RaiseEvent eHeightChanged(Me, EventArgs.Empty)
		End Sub

		Private Function CalcHeight(ByVal cache As GraphicsCache) As Integer Implements IAutoHeightControl.CalcHeight
			If ViewInfo.IsReady Then
				Dim ih As IHeightAdaptable = TryCast(ViewInfo, IHeightAdaptable)
				If ih IsNot Nothing Then
					Return ih.CalcHeight(cache, Width)
				End If
			End If
			Return Height

		End Function
		#End Region
		Shared Sub New()
			CustomRepositoryItem.Register()
		End Sub
		Public Sub New()
			MyBase.New()
			AddHandler Properties.ControlTypeChanged, AddressOf Properties_ControlTypeChanged
			UpdateControls()
			'this.Dock = System.Windows.Forms.DockStyle.Fill;
		End Sub

		Private Sub Properties_ControlTypeChanged(ByVal sender As Object, ByVal e As EventArgs)
			UpdateControls()
		End Sub
		Private Sub UpdateControls()
			Controls.Clear()
			If Properties.ControlType Is Nothing Then
				Return
			End If
			Controls.Add(Properties.EditorControl)
			AddHandler TryCast(Properties.EditorControl, IEditValue).EditValueChanged, AddressOf editor_EditValueChanged
			TryCast(Properties.EditorControl, IEditValue).EditValue = EditValue
		End Sub

		Private Sub editor_EditValueChanged(ByVal sender As Object, ByVal e As EventArgs)

			EditValue = (TryCast(Properties.EditorControl, IEditValue)).EditValue
			IsModified = True
		End Sub
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return CustomRepositoryItem.EditorName
			End Get
		End Property
		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
		Public Shadows ReadOnly Property Properties() As CustomRepositoryItem
			Get
				Return TryCast(MyBase.Properties, CustomRepositoryItem)
			End Get
		End Property
		Protected Overrides Sub OnPropertiesChanged()
			MyBase.OnPropertiesChanged()
			Me.RaiseHeightChanged()
		End Sub
		Public Overrides Property EditValue() As Object
			Get
				Return MyBase.EditValue
			End Get
			Set(ByVal value As Object)
				MyBase.EditValue = value
				TryCast(Properties.EditorControl, IEditValue).EditValue = EditValue
			End Set
		End Property
	End Class

End Namespace

File: CS/CustomControlInGrid/CustomRepositoryItem/CustomControl.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Controls;
using System.ComponentModel;
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors.ViewInfo;

namespace CustomControlInGrid
{
    class CustomControl : BaseEdit, IAutoHeightControl
    {
        #region IAutoHeightControl implement
        bool IAutoHeightControl.SupportsAutoHeight { get { return true; } }

        public event EventHandler eHeightChanged;
        event EventHandler IAutoHeightControl.HeightChanged
        {
            add { eHeightChanged += value; }
            remove { eHeightChanged -= value; }
        }
        protected void RaiseHeightChanged()
        {
            if (eHeightChanged != null)
                eHeightChanged(this, EventArgs.Empty);
        }

        int IAutoHeightControl.CalcHeight(GraphicsCache cache)
        {
            if (ViewInfo.IsReady)
            {
                IHeightAdaptable ih = ViewInfo as IHeightAdaptable;
                if (ih != null) return ih.CalcHeight(cache, Width);
            }
            return Height;

        }
        #endregion
        static CustomControl()
        {
            CustomRepositoryItem.Register();
        }
        public CustomControl()
            : base()
        {
            Properties.ControlTypeChanged += new EventHandler(Properties_ControlTypeChanged);
            UpdateControls();
            //this.Dock = System.Windows.Forms.DockStyle.Fill;
        }

        void Properties_ControlTypeChanged(object sender, EventArgs e)
        {
            UpdateControls();
        }
        void UpdateControls()
        {
            Controls.Clear();
            if (Properties.ControlType == null)
                return;
            Controls.Add(Properties.EditorControl);
            (Properties.EditorControl as IEditValue).EditValueChanged += new EventHandler(this.editor_EditValueChanged);
            (Properties.EditorControl as IEditValue).EditValue = EditValue;
        }

        void editor_EditValueChanged(object sender, EventArgs e)
        {

            EditValue = (Properties.EditorControl as IEditValue).EditValue;
            IsModified = true;
        }
        public override string EditorTypeName
        {
            get { return CustomRepositoryItem.EditorName; }
        }
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new CustomRepositoryItem Properties
        {
            get { return base.Properties as CustomRepositoryItem; }
        }
        protected override void OnPropertiesChanged()
        {
            base.OnPropertiesChanged();
            this.RaiseHeightChanged();
        }
        public override object EditValue
        {
            get
            {
                return base.EditValue;
            }
            set
            {
                base.EditValue = value;
                (Properties.EditorControl as IEditValue).EditValue = EditValue;
            }
        }
    }

}
